### ISSUE TITLE
[RFC]: Support Speculative Decoding with Decode Context Parallelism

### ISSUE THREAD

**Issue by pisceskkk** [vllm-project/vllm#50391]
### Motivation.

vLLM currently does not support using speculative decoding together with Decode Context Parallelism (DCP). This limitation applies to all DCP configurations, including both `cp_kv_cache_interleave_size = 1` and larger interleave sizes.

DCP shards KV-cache tokens across ranks according to their global positions. 
For a global sequence length `L`, DCP size `D`, interleave size `I`, and rank `r`, the rank-local KV length is:

```text
local_len(r, L) =
    floor(L / (D * I)) * I
    + clamp(L % (D * I) - r * I, 0, I)
```

During speculative verification, one request contains multiple query tokens. Query token `j`, with historical context length `H`, must see:

```text
local_len(r, H + j + 1)
```

KV tokens on rank `r`. These boundaries differ across queries and ranks and cannot in general be inferred from the final rank-local KV length.

DCP also affects the proposer. After speculative acceptance or rejection, the global sequence length changes, but the corresponding rank-local sequence length cannot be updated by directly adding or subtracting the number of accepted or rejected tokens. It must be derived again from the corrected global length and DCP topology.

The goal of this RFC is to add complete DCP and speculative-decoding compatibility for:

1. autoregressive speculative proposers;
2. multi-token target verification in attention backends;
3. different context lengths, DCP ranks, interleave sizes, and acceptance
   results;
4. eager execution and CUDA Graph replay.

### 1. Speculative decoding proposer

The proposer must generate DCP-aware attention metadata for every autoregressive draft step.

After speculative rejection is applied, the proposer has a corrected global sequence length. Before building the next draft attention metadata, it must recompute the corresponding rank-local sequence length from:

```text
global sequence length
DCP size
DCP rank
KV-cache interleave size
```

The proposer must not update rank-local length using:

```python
local_seq_len += 1
local_seq_len -= num_rejected
```

Consecutive global tokens may belong to different DCP ranks, including when `cp_kv_cache_interleave_size = 1`.

#### Draft metadata update

Before building attention metadata for every draft decode step, the proposer should populate a persistent DCP local-length buffer:

```python
prepare_dcp_local_seq_lens(
    input_buffers.dcp_local_seq_lens,
    input_buffers.seq_lens,
    num_reqs,
    dcp_size,
    dcp_rank,
    cp_interleave,
)
```

The resulting tensor must be passed to the attention metadata builder:

```python
build_attn_metadata(
    ...,
    seq_lens=input_buffers.seq_lens,
    dcp_local_seq_lens=input_buffers.dcp_local_seq_lens,
)
```

The draft-step metadata flow is:

```text
Apply rejection to global sequence length
        ↓
Prepare the absolute draft position
        ↓
Prepare DCP-aware slot mapping
        ↓
Recompute rank-local sequence length
        ↓
Build draft attention metadata
        ↓
Run proposer attention
        ↓
Advance global state for the next draft step
```

Rank-local sequence length must be recomputed after every global sequence-length change.

#### Problem

For a target verification request with historical length `H` and `N` query tokens, the final local KV length on rank `r` is:

```text
local_len(r, H + N)
```

However, query token `j` may only attend to:

```text
local_len(r, H + j + 1)
```

local KV tokens.

A conventional causal attention kernel generally derives all query boundaries from the final KV length. Under DCP, this may expose future speculative KV or hide valid KV, producing incorrect verifier logits and acceptance decisions.

For example:
<details>
<summary><b>Example: DCP=2, MTP=3, context length=5</b></summary>

Assume:

- `cp_kv_cache_interleave_size = 1`
- MTP3 produces a target decode query length of 4
- `Q0–Q3` are at global positions `5–8`
- `1` means visible and `0` means masked

The global KV ownership is:

| Global position | 0 | 1 | 2 | 3 | 4 | 5 (Q0) | 6 (Q1) | 7 (Q2) | 8 (Q3) |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| DCP owner | R0 | R1 | R0 | R1 | R0 | R1 | R0 | R1 | R0 |

#### Rank 0

The final local KV sequence is `[k0, k2, k4, k6, k8]`.

| Query | k0 | k2 | k4 | k6 | k8 | Visible local KV length |
|---|---:|---:|---:|---:|---:|---:|
| Q0, global position 5 | 1 | 1 | 1 | 0 | 0 | 3 |
| Q1, global position 6 | 1 | 1 | 1 | 1 | 0 | 4 |
| Q2, global position 7 | 1 | 1 | 1 | 1 | 0 | 4 |
| Q3, global position 8 | 1 | 1 | 1 | 1 | 1 | 5 |

#### Rank 1

The final local KV sequence is `[k1, k3, k5, k7]`.

| Query | k1 | k3 | k5 | k7 | Visible local KV length |
|---|---:|---:|---:|---:|---:|
| Q0, global position 5 | 1 | 1 | 1 | 0 | 3 |
| Q1, global position 6 | 1 | 1 | 1 | 0 | 3 |
| Q2, global position 7 | 1 | 1 | 1 | 1 | 4 |
| Q3, global position 8 | 1 | 1 | 1 | 1 | 4 |

A conventional bottom-right causal mask derives visibility only from the local KV length and query length. It would produce:

| | Rank 0 | Rank 1 |
|---|---|---|
| Correct per-query visibility | `[3, 4, 4, 5]` | `[3, 3, 4, 4]` |
| Bottom-right causal visibility | `[2, 3, 4, 5]` | `[1, 2, 3, 4]` |

The mismatch occurs because advancing one global query token does not necessarily increase the local KV length on every DCP rank. Therefore, the backend needs per-query rank-local KV boundaries, a rank-specific mask, or a context/current split implementation.

</details>

The following strategies can implement the required per-query rank-local visibility.

#### Strategy A: Direct CP-aware attention

The attention kernel receives the information required to convert global causal boundaries into rank-local KV boundaries:

```text
DCP size
DCP rank
interleave size
global query positions
global sequence lengths
```

For FlashAttention, this strategy can be used when the selected kernel interface supports CP rank, CP world size, and KV-cache interleave.

Flow:

```text
Q[N] + rank-local paged KV
        +
global positions and DCP topology
        ↓
CP-aware attention
        ↓
partial output + LSE
        ↓
DCP output/LSE reduction
```

Performance impact:

- one main attention call;
- no custom dense mask;
- no virtual batch expansion;
- lowest dispatch and metadata overhead;
- requires backend-specific kernel support;
- every supported interleave configuration must be validated.

This provides the highest expected performance but has the highest kernel implementation cost.

#### Strategy B: Split context and current KV

Historical KV and the current speculative KV block are computed separately.

Historical path:

```text
Q[N] × rank-local paged historical KV
non-causal
```

Current path:

```text
Q[N] × current K/V[N]
causal
```

The two attention states are merged using their output and LSE. This can be implemented using a chunked-prefill-style flow: the paged cache represents historical context, while raw current K/V represents the speculative token block.

Flow:

```text
Q × local paged history
        ↓
history output + LSE
                         \
                          → LSE/state merge → DCP reduction
                         /
Q × current K/V
        ↓
causal output + LSE
```

Performance impact:

- supports arbitrary DCP ranks and interleave layouts;
- does not require a rank-specific dense mask;
- requires two attention calls and one state merge;
- introduces additional workspace and launch overhead;
- MLA implementations may require context chunk gathering or reconstruction.

This provides a portable implementation when direct CP-aware attention is not available.

#### Strategy C: Custom mask

The backend constructs a rank-specific attention mask that explicitly describes the visible local KV range for every speculative query.

For query token `j` on DCP rank `r`:

```text
local_boundary[j] = local_len(r, H + j + 1)
mask[j, k] = k >= local_boundary[j]
```

Each DCP rank generates a different mask because its local KV length and per-query visibility boundaries are different.

Flow:

```text
Global history and query positions
        ↓
Compute rank-local boundary for every query
        ↓
Construct rank-specific custom mask
        ↓
Q[N] × rank-local paged KV
        ↓
partial output + LSE
        ↓
DCP output/LSE reduction
```

Performance impact:

- supports arbitrary DCP sizes, interleave configurations, and variable query
  lengths;
- retains a single main attention call;
- requires `O(batch × N × local_KV)` mask memory;
- introduces device-side or host-to-device mask construction overhead;
- increases CUDA Graph input and persistent-buffer sizes;
- mask generation cost grows with context length;
- **may disable optimized causal kernels and force the backend to use a more
  generic masked-attention kernel**.

This is suitable for backends with efficient custom-mask support.

#### Strategy D: Expand N tokens into N virtual requests

One request containing `N` speculative query tokens is represented as `N` virtual single-query attention rows.

Virtual row `j` uses:

```text
query position         = H + j
global sequence length = H + j + 1
local sequence length  = local_len(r, H + j + 1)
query length           = 1
```

Each virtual row reuses the original request's KV-cache block table while providing an independent context length.

Flow:

```text
One request with N query tokens
        ↓
N virtual q_len=1 rows
        ↓
Duplicate block-table references and sequence metadata
        ↓
Batched single-token DCP decode
        ↓
Restore the original token order
```

Performance impact:

- reuses single-token DCP decode kernels;
- supports arbitrary DCP layouts without a custom mask;
- expands the logical batch size by up to `N`;
- duplicates block-table and sequence metadata;
- increases scheduling and metadata-processing overhead;
- reduces batching efficiency;
- requires larger CUDA Graph batch dimensions.

This provides a correctness-oriented fallback but is expected to have the largest overhead as `N` grows.

#### Strategy comparison

| Strategy | Attention calls | Extra metadata | Kernel changes | Expected performance |
|---|---:|---:|---:|---|
| Direct CP-aware | 1 | Low | High | Best |
| Context/current split | 2 + merge | Low | Medium | Good and portable |
| Custom mask | 1 | `O(B×N×local_KV)` | Low/Medium | Context and kernel dependent |
| Virtual single-query rows | 1 expanded batch | Block-table duplication | Low/Medium | Lower for larger `N` |

### CC List.

@LucasWilkinson @GirasoleY @ZJY0516 

### Any Other Things.

Related Links
1. PR #39632 introduces DCP-aware attention metadata reuse across MTP draft steps. It is related to the proposer changes in this RFC, while this RFC additionally defines how context lengths must be updated consistently across draft iterations and acceptance outcomes.
2. Issue #45425 reports incorrect results when a backend without DCP variable-length decode support executes MTP verification under full or piecewise CUDA graphs. It demonstrates why this RFC requires explicit backend handling of per-query rank-local KV boundaries instead of relying on conventional multi-token decode metadata.
3. PR #45426 computes per-token rank-local causal boundaries for the FlashMLA sparse path and enables DCP variable-length decode when the KV-cache interleave size is one. It provides a backend-specific implementation of the local-boundary approach considered by this RFC, while this RFC targets a backend-neutral design without assuming a particular interleave size.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by HaizhouPeng** [comment]
++++

**Comment by Mahnoor-Zaffar** [comment]
I'm excited to tackle the compatibility issue between speculative decoding and DCP! I'll refactor the local sequence length calculation to ensure it derives from the global sequence length and DCP parameters after each speculative acceptance or rejection. This should resolve the current limitations. Could you please assign this issue to me?

**Comment by NickLucche** [comment]
cc @benchislett
